cider/tests/utils/DuckDbQueryRunner.cpp

```diff
@@ -388,6 +390,164 @@ void addColumnDataToCiderBatch<CiderByteArray>(
+  for (int j = 0; j < row_num; j++) {
+    if (dataChunk->GetValue(i, j).IsNull()) {
+      std::memset(data_buffer + j * sizeof(CiderByteArray), 0, sizeof(CiderByteArray));
+      CiderBitUtils::clearBitAt(null_buffer, j);
+      ++null_count;
+    } else {
+      std::string value = dataChunk->GetValue(i, j).GetValue<std::string>();
+      // memory leak risk. user should manually free.
+      uint8_t* value_buf =
+          reinterpret_cast<uint8_t*>(allocator->allocate(value.length()));
+      std::memcpy(value_buf, value.c_str(), value.length());
+      buf[j].len = value.length();
+      buf[j].ptr = value_buf;
+      CiderBitUtils::setBitAt(null_buffer, j);
+    }
+  }
+  child->setNullCount(null_count);
+}
+
+CiderBatch DuckDbResultConvertor::fetchOneArrowFormattedBatch(
+    std::unique_ptr<duckdb::DataChunk>& chunk) {
+  // Construct ArrowSchema
+  // First build a CiderTableSchema and then convert it
+  std::vector<::substrait::Type> batch_types;
+  std::vector<std::string> batch_names;
+
+  int col_num = chunk->ColumnCount();
+  int row_num = chunk->size();
+  std::vector<::duckdb::LogicalType> types = chunk->GetTypes();
+  for (int i = 0; i < col_num; ++i) {
+    auto type = types[i];
+    batch_types.emplace_back(std::move(convertToSubstraitType(type)));
+    batch_names.emplace_back("");
+    if (!isDuckDb2CiderBatchSupportType(type)) {
+      CIDER_THROW(CiderCompileException,
+                  "Non numberic type " + type.ToString() + " not supported yet");
+    }
+  }
+
+  auto cider_schema = std::make_shared<CiderTableSchema>(batch_names, batch_types, "");
+
+  /// NOTE: (YBRua) This function currently only support primitive scalar types
+  /// It cannot convert Date/Time/Varchar to ArrowSchema
+  auto arrow_schema =
+      CiderBatchUtils::convertCiderTableSchemaToArrowSchema(*cider_schema);
+
+  // Construct ArrowArray
+  /// NOTE: (YBRua) DuckDb provides an API ToArrowArray(), but we are NOT using it
+  /// since the structs for DATE and VARCHAR in Cider differ from those used in DuckDB
+  auto batch =
+      StructBatch::Create(arrow_schema, std::make_shared<CiderDefaultAllocator>());
+  CHECK(batch->resizeBatch(row_num, true));
+  for (int i = 0; i < col_num; ++i) {
+    auto child = batch->getChildAt(i);
+    auto type = types[i];
+    switch (type.id()) {
+      case ::duckdb::LogicalTypeId::HUGEINT:
+      case ::duckdb::LogicalTypeId::BIGINT:
+      case ::duckdb::LogicalTypeId::TIMESTAMP:
+      case ::duckdb::LogicalTypeId::TIME:
+        addColumnDataToScalarBatch<int64_t>(chunk, i, row_num, child.get());
+        break;
+      case ::duckdb::LogicalTypeId::INTEGER:
+        addColumnDataToScalarBatch<int32_t>(chunk, i, row_num, child.get());
+        break;
+      case ::duckdb::LogicalTypeId::SMALLINT:
+        addColumnDataToScalarBatch<int16_t>(chunk, i, row_num, child.get());
+        break;
+      case ::duckdb::LogicalTypeId::BOOLEAN:
+        addColumnDataToScalarBatch<bool>(chunk, i, row_num, child.get());
```

Comment:
may be put these in a order from small to large, eg bool, tiny, small, int....

Reply:
These lines have been removed. We are using methods provided by DuckDb to directly construct an ArrowArray.